### ISSUE TITLE
Kernel module compilation does not clean leftovers properly on Fedora 22

### ISSUE THREAD

**Issue by danielkza** [zfsonlinux/zfs#3538]
After upgrading ZoL's package on Fedora 22 today, `dkms status` ended up broken afterwards with multiple "file not found" messages mixed with the actual output. Context of what happened:
- Previous ZoL version was 0.6.4.1, attempting to upgrade to 0.6.4.2 using DNF
- A kernel upgrade from 4.0.4 to 4.0.5 was part of the same transaction
- Kernel 3.19.7 was being removed as part of the transaction, but the ZFS upgrade built for it anyway. I suspect it ran before the kernel upgrade, built for the 3 older kernels, then the 4.0.5 installation triggered the build just for itself
- Afterwards, `/var/lib/dkms/0.6.4.1` still existed, with a broken symlink named `source` inside.
- It seems it was caused by DKMS believing 3.19.7 is still active, and it having it's module as version 0.6.4.1, which no longer exists.
### System Information
- Fedora 22 x86-64
- dkms-2.2.0.3-30.git.7c3e7c5.fc22
- zfs-0.6.4.1-1.fc22 (before upgrade)
- zfs-0.6.4.2-1.fc22 (after upgrade)
### Scriptlet outputs for transaction

```
Loading new spl-0.6.4.2 DKMS files...
dpkg: warning: version '3.19.7-200.fc21.x86_64' has bad syntax: invalid character in revision number
dpkg: warning: version '3.18.7-200.fc21.x86_64
3.19.4' has bad syntax: invalid character in revision number
dpkg: warning: version '4.0.4-301.fc22.x86_64' has bad syntax: invalid character in revision number
dpkg: warning: version '3.19.7-200.fc21.x86_64' has bad syntax: invalid character in revision number
dpkg: warning: version '4.0.4-303.fc22.x86_64' has bad syntax: invalid character in revision number
dpkg: warning: version '4.0.4-301.fc22.x86_64' has bad syntax: invalid character in revision number
dpkg: warning: version '4.0.5-300.fc22.x86_64' has bad syntax: invalid character in revision number
dpkg: warning: version '4.0.4-303.fc22.x86_64' has bad syntax: invalid character in revision number
Building for 3.18.7-200.fc21.x86_64, 3.19.4-200.fc21.x86_64, 3.19.5-200.fc21.x86_64, 3.19.7-200.fc21.x86_64, 4.0.4-301.fc22.x86_64, 4.0.4-303.fc22.x86_64 and 4.0.5-300.fc22.x86_64
Module build for kernel 3.18.7-200.fc21.x86_64 was skipped since the
kernel source for this kernel does not seem to be installed.
Module build for kernel 3.19.4-200.fc21.x86_64 was skipped since the
kernel source for this kernel does not seem to be installed.
Module build for kernel 3.19.5-200.fc21.x86_64 was skipped since the
kernel source for this kernel does not seem to be installed.
Building initial module for 3.19.7-200.fc21.x86_64
Done.

spl:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for spl.ko
exactly matches what is already found in kernel 3.19.7-200.fc21.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

splat.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for splat.ko
exactly matches what is already found in kernel 3.19.7-200.fc21.x86_64.
DKMS will not replace this module.
You may override by specifying --force.
Adding any weak-modules

depmod....

DKMS: install completed.
Building initial module for 4.0.4-301.fc22.x86_64
Done.

spl:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for spl.ko
exactly matches what is already found in kernel 4.0.4-301.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

splat.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for splat.ko
exactly matches what is already found in kernel 4.0.4-301.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.
Adding any weak-modules

depmod...

DKMS: install completed.
Building initial module for 4.0.4-303.fc22.x86_64
Done.

spl:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for spl.ko
exactly matches what is already found in kernel 4.0.4-303.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

splat.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for splat.ko
exactly matches what is already found in kernel 4.0.4-303.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.
Adding any weak-modules

depmod...

DKMS: install completed.
Building initial module for 4.0.5-300.fc22.x86_64
Done.

spl:
Running module version sanity check.
 - Original module
   - No original module exists within this kernel
 - Installation
   - Installing to /lib/modules/4.0.5-300.fc22.x86_64/extra/

splat.ko:
Running module version sanity check.
 - Original module
   - No original module exists within this kernel
 - Installation
   - Installing to /lib/modules/4.0.5-300.fc22.x86_64/extra/
Adding any weak-modules

depmod...

DKMS: install completed.
Loading new zfs-0.6.4.2 DKMS files...
dpkg: warning: version '3.19.7-200.fc21.x86_64' has bad syntax: invalid character in revision number
dpkg: warning: version '3.18.7-200.fc21.x86_64
3.19.4' has bad syntax: invalid character in revision number
dpkg: warning: version '4.0.4-301.fc22.x86_64' has bad syntax: invalid character in revision number
dpkg: warning: version '3.19.7-200.fc21.x86_64' has bad syntax: invalid character in revision number
dpkg: warning: version '4.0.4-303.fc22.x86_64' has bad syntax: invalid character in revision number
dpkg: warning: version '4.0.4-301.fc22.x86_64' has bad syntax: invalid character in revision number
dpkg: warning: version '4.0.5-300.fc22.x86_64' has bad syntax: invalid character in revision number
dpkg: warning: version '4.0.4-303.fc22.x86_64' has bad syntax: invalid character in revision number
Building for 3.18.7-200.fc21.x86_64, 3.19.4-200.fc21.x86_64, 3.19.5-200.fc21.x86_64, 3.19.7-200.fc21.x86_64, 4.0.4-301.fc22.x86_64, 4.0.4-303.fc22.x86_64 and 4.0.5-300.fc22.x86_64
Module build for kernel 3.18.7-200.fc21.x86_64 was skipped since the
kernel source for this kernel does not seem to be installed.
Module build for kernel 3.19.4-200.fc21.x86_64 was skipped since the
kernel source for this kernel does not seem to be installed.
Module build for kernel 3.19.5-200.fc21.x86_64 was skipped since the
kernel source for this kernel does not seem to be installed.
Building initial module for 3.19.7-200.fc21.x86_64
Done.

zavl:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for zavl.ko
exactly matches what is already found in kernel 3.19.7-200.fc21.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

znvpair.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for znvpair.ko
exactly matches what is already found in kernel 3.19.7-200.fc21.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

zunicode.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for zunicode.ko
exactly matches what is already found in kernel 3.19.7-200.fc21.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

zcommon.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for zcommon.ko
exactly matches what is already found in kernel 3.19.7-200.fc21.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

zfs.ko:
Running module version sanity check.
 - Original module
   - This kernel never originally had a module by this name
 - Installation
   - Installing to /lib/modules/3.19.7-200.fc21.x86_64/extra/

zpios.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for zpios.ko
exactly matches what is already found in kernel 3.19.7-200.fc21.x86_64.
DKMS will not replace this module.
You may override by specifying --force.
Adding any weak-modules

depmod...

DKMS: install completed.
Building initial module for 4.0.4-301.fc22.x86_64
Done.

zavl:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for zavl.ko
exactly matches what is already found in kernel 4.0.4-301.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

znvpair.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for znvpair.ko
exactly matches what is already found in kernel 4.0.4-301.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

zunicode.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for zunicode.ko
exactly matches what is already found in kernel 4.0.4-301.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

zcommon.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for zcommon.ko
exactly matches what is already found in kernel 4.0.4-301.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

zfs.ko:
Running module version sanity check.
 - Original module
   - This kernel never originally had a module by this name
 - Installation
   - Installing to /lib/modules/4.0.4-301.fc22.x86_64/extra/

zpios.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for zpios.ko
exactly matches what is already found in kernel 4.0.4-301.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.
Adding any weak-modules

depmod...

DKMS: install completed.
Building initial module for 4.0.4-303.fc22.x86_64
Done.

zavl:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for zavl.ko
exactly matches what is already found in kernel 4.0.4-303.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

znvpair.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for znvpair.ko
exactly matches what is already found in kernel 4.0.4-303.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

zunicode.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for zunicode.ko
exactly matches what is already found in kernel 4.0.4-303.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

zcommon.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for zcommon.ko
exactly matches what is already found in kernel 4.0.4-303.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.

zfs.ko:
Running module version sanity check.
 - Original module
   - This kernel never originally had a module by this name
 - Installation
   - Installing to /lib/modules/4.0.4-303.fc22.x86_64/extra/

zpios.ko:
Running module version sanity check.

Good news! Module version 0.6.4.2-1 for zpios.ko
exactly matches what is already found in kernel 4.0.4-303.fc22.x86_64.
DKMS will not replace this module.
You may override by specifying --force.
Adding any weak-modules

depmod...

DKMS: install completed.
Building initial module for 4.0.5-300.fc22.x86_64
Done.

zavl:
Running module version sanity check.
 - Original module
   - No original module exists within this kernel
 - Installation
   - Installing to /lib/modules/4.0.5-300.fc22.x86_64/extra/

znvpair.ko:
Running module version sanity check.
 - Original module
   - No original module exists within this kernel
 - Installation
   - Installing to /lib/modules/4.0.5-300.fc22.x86_64/extra/

zunicode.ko:
Running module version sanity check.
 - Original module
   - No original module exists within this kernel
 - Installation
   - Installing to /lib/modules/4.0.5-300.fc22.x86_64/extra/

zcommon.ko:
Running module version sanity check.
 - Original module
   - No original module exists within this kernel
 - Installation
   - Installing to /lib/modules/4.0.5-300.fc22.x86_64/extra/

zfs.ko:
Running module version sanity check.
 - Original module
   - No original module exists within this kernel
 - Installation
   - Installing to /lib/modules/4.0.5-300.fc22.x86_64/extra/

zpios.ko:
Running module version sanity check.
 - Original module
   - No original module exists within this kernel
 - Installation
   - Installing to /lib/modules/4.0.5-300.fc22.x86_64/extra/
Adding any weak-modules

depmod...

DKMS: install completed.
warning: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.45-40.b14.fc22.x86_64/jre/lib/security/US_export_policy.jar created as /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.45-40.b14.fc22.x86_64/jre/lib/security/US_export_policy.jar.rpmnew
warning: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.45-40.b14.fc22.x86_64/jre/lib/security/java.security created as /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.45-40.b14.fc22.x86_64/jre/lib/security/java.security.rpmnew
warning: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.45-40.b14.fc22.x86_64/jre/lib/security/local_policy.jar created as /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.45-40.b14.fc22.x86_64/jre/lib/security/local_policy.jar.rpmnew
Redirecting to /bin/systemctl start  atd.service
dkms: removing: spl 0.6.4.2 (3.19.7-200.fc21.x86_64) (x86_64)

-------- Uninstall Beginning --------
Module:  spl
Version: 0.6.4.2
Kernel:  3.19.7-200.fc21.x86_64 (x86_64)
-------------------------------------

Status: Before uninstall, this module version was ACTIVE on this kernel.
Removing any linked weak-modules

spl.ko:
 - Uninstallation
   - Deleting from: /lib/modules/3.19.7-200.fc21.x86_64/
rmdir: failed to remove ‘’: No such file or directory
 - Original module
   - No original module was found for this module on this kernel.
   - Use the dkms install command to reinstall any previous module version.


splat.ko:
 - Uninstallation
   - Deleting from: /lib/modules/3.19.7-200.fc21.x86_64/
rmdir: failed to remove ‘’: No such file or directory
 - Original module
   - No original module was found for this module on this kernel.
   - Use the dkms install command to reinstall any previous module version.

depmod...

DKMS: uninstall completed.
grubby fatal error: unable to find a suitable template
grubby: doing this would leave no kernel entries. Not writing out new config.
warning: file /lib/modules/3.19.7-200.fc21.x86_64/updates: remove failed: No such file or directory
warning: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.45-39.b14.fc22.x86_64/jre/lib/security/local_policy.jar saved as /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.45-39.b14.fc22.x86_64/jre/lib/security/local_policy.jar.rpmsave
warning: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.45-39.b14.fc22.x86_64/jre/lib/security/java.security saved as /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.45-39.b14.fc22.x86_64/jre/lib/security/java.security.rpmsave
warning: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.45-39.b14.fc22.x86_64/jre/lib/security/US_export_policy.jar saved as /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.45-39.b14.fc22.x86_64/jre/lib/security/US_export_policy.jar.rpmsave
grubby fatal error: unable to find a suitable template
grubby fatal error: unable to find a suitable template
grubby: doing this would leave no kernel entries. Not writing out new config.
Error! Could not locate dkms.conf file.
File:  does not exist.
```

**Comment by tjikkun** [comment]
I see the same on CloudLinux 6.6:

### System Information
* CloudLinux 6.6 x86-64
* dkms-2.2.0.3-30.git.7c3e7c5.el6.noarch
* zfs-0.6.4.1-1.el6.x86_64 (before upgrade)
* zfs-0.6.4.2-1.el6.x86_64 (after upgrade)

**Comment by Lalufu** [comment]
This happens because the old version source trees are not installed any more, but the old version DKMS build directories (/var/lib/dkms/spl/0.6.4.1 for example) still contain a symlink (source) to them. This symlink is now broken, and this seems to confuse DKMS to no end.

It seems to be sufficient to just remove the symlink file to make DKMS happy again. This could be done in the %post section of the RPM files.

**Comment by behlendorf** [comment]
Closing.  The spl and zfs modules have been moved in to a single dkms package to avoid issues such as this.

**Comment by jochendemuth** [comment]
Although combining the zfs modules into a single dkms package is much appreciated, it does nothing to fix this error. It still occurs with every new version of zfs. 
Observed latest in Fedora 30 during the upgrade from zfs 0.8.0 to zfs 0.8.1.
As part of the package upgrade, the old source code is removed, leaving a stale symlinks in the /var/lib/dkms/zfs tree. While not nice in general, it only is a problem (as correctly pointed out in earlier comments) because the old configuration file is removed invalidating the dkms configuration.

# dkms status
Error! Could not locate dkms.conf file.
File: /var/lib/dkms/zfs/0.8.0/source/dkms.conf does not exist.
#

I think the correct fix would be to add a 'dkms remove zfs/0.8.0 --all' command to the upgrade/cleanup script of the zfs package - **before** the source code of the old version gets removed.

**Comment by Lalufu** [comment]
F30 has other, unrelated DKMS issues. See #8763.
